src/coap_io.c

```diff
@@ -850,9 +850,10 @@ coap_network_read(coap_socket_t *sock, coap_packet_t *packet) {
 #else
       if (errno == ECONNREFUSED) {
 #endif
-	/* client-side ICMP destination unreachable, ignore it */
-	coap_log(LOG_WARNING, "coap_network_read: unreachable\n");
-	return -2;
+        /* client-side ICMP destination unreachable, ignore it */
```

Comment:
I hear you re ICMP Not reachable.  The application requires one socket per target server, and as DTLS is never established, the existing socket cannot continue to be used.  We have to close the existing session and start a new coap_new_client_session_psk() to start the DTLS negotiations again.  I will remove the coap_socket_close() from this change and see how it can be done from within the application.

Reply:
However, as the return of -2 from coap_network_read() triggers a call to coap_session_reset() which not only resets any pending traffic, but also frees off the TLS session in coap_dtls_free_session(), so the session is unusable for any further TLS traffic -not sure if this is the intention.

Reply:
Yes it is the intention. A new DTLS session will automatically be established when a new message is queued for sending. At least this is what it is supposed to do.

Reply:
Our code is now correctly freeing off the session before starting the next one and no fd leeks found so far.  It was my intention to test out the auto re-establishment today, but I did not have time.